lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,298 @@
+                          &block) == control_deps_map[blockInPath].end())
+              control_deps_map[blockInPath].push_back(&block);
+          }
+      }
+    }
+  }
+
+  // Up to this point, we have correct direct dependencies that do not include
+  // nested dependencies (i.e., dependencies of a block's dependencies) call the
+  // following function to include nested dependencies
+  addDepsOfDeps(funcOp, control_deps_map);
+  all_control_deps_maps.emplace_back(control_deps_map);
+
+  // Extract the forward dependencies for this function
+  identifyForwardControlDeps(funcOp);
+}
+
+void ControlDependenceAnalysis::enumeratePathsInPostDomTree(
+    Block *start_block, Block *end_block, Region *funcReg,
+    llvm::DominatorTreeBase<Block, true> *postDomTree,
+    SmallVector<PathInDomTree, 4> *traversed_nodes) {
+
+  DominanceInfoNode *start_node = postDomTree->getNode(start_block);
+  DominanceInfoNode *end_node = postDomTree->getNode(end_block);
+
+  DenseMap<DominanceInfoNode *, bool> is_visited;
+  for (Block &block : funcReg->getBlocks())
+    is_visited[postDomTree->getNode(&block)] = false;
+
+  PathInDomTree path;
+  path.resize(funcReg->getBlocks().size());
+  int path_index = 0;
+
+  enumeratePathsInPostDomTreeUtil(start_node, end_node, is_visited, path,
+                                  path_index, traversed_nodes);
+}
+
+// DFS to return all nodes in the path between the start_node and end_node (not
+// including start_node and end_node) in the postDom tree
+void ControlDependenceAnalysis::enumeratePathsInPostDomTreeUtil(
+    DominanceInfoNode *start_node, DominanceInfoNode *end_node,
+    DenseMap<DominanceInfoNode *, bool> is_visited, PathInDomTree path,
+    int path_index, SmallVector<PathInDomTree, 4> *traversed_nodes) {
+  is_visited[start_node] = true;
+  path[path_index] = start_node;
+  path_index++;
+
+  // if start is same as end, we have completed one path so push it to
+  // traversed_nodes
+  if (start_node == end_node) {
+    // slice of the path from its beginning until the path_index
+    PathInDomTree actual_path;
+    for (auto i = 0; i < path_index; i++)
+      actual_path.push_back(path[i]);
+    traversed_nodes->push_back(actual_path);
+
+  } else {
+    // loop over the children of start_node
+    for (DominanceInfoNode::iterator iter = start_node->begin();
+         iter < start_node->end(); iter++) {
+      if (!is_visited[*iter]) {
+        enumeratePathsInPostDomTreeUtil(*iter, end_node, is_visited, path,
+                                        path_index, traversed_nodes);
+      }
+    }
+  }
+
+  // remove this node from path and mark it as unvisited
+  path_index--;
+  is_visited[start_node] = false;
+}
+
+void ControlDependenceAnalysis::addDepsOfDeps(
+    FuncOp &funcOp,
+    DenseMap<Block *, SmallVector<Block *, 4>> &control_deps_map) {
+  Region &funcReg = funcOp.getRegion();
+
+  for (Block &block : funcReg.getBlocks()) {
+    SmallVector<Block *, 4> block_deps = control_deps_map[&block];
```

Comment:
This unnecessarily copies the entire vector do `SmallVector<Block *, 4>& block_deps = control_deps_map[&block];` to use a reference instead.